crates/ra_assists/src/assists/early_return.rs

```diff
@@ -0,0 +1,185 @@
+use hir::db::HirDatabase;
+use itertools::Itertools;
+use ra_syntax::{ast::Block, ast::FnDef, ast::IfExpr, ast::LoopExpr, ast::ReturnExpr, AstNode};
+
+use crate::{Assist, AssistCtx, AssistId};
+
+pub(crate) fn convert_to_guarded_return(mut ctx: AssistCtx<impl HirDatabase>) -> Option<Assist> {
+    let if_expr: IfExpr = ctx.node_at_offset()?;
+    let expr = if_expr.condition()?.expr()?;
+    let then_block = if_expr.then_branch()?.block()?;
+    if if_expr.else_branch().is_some() {
+        return None;
+    }
+
+    let parent_block = if_expr.syntax().parent()?.ancestors().find_map(Block::cast)?;
+
+    if parent_block.syntax().last_child()? != *if_expr.syntax() {
+        return None;
+    }
+
+    let first_in_then_block = then_block.syntax().first_child()?.clone();
+    if ReturnExpr::can_cast(first_in_then_block.kind())
+        || first_in_then_block.children().any(|x| ReturnExpr::can_cast(x.kind()))
+    {
+        return None;
+    }
+
+    let parent_container = parent_block.syntax().parent()?.parent()?;
+    let early_expression = if LoopExpr::can_cast(parent_container.kind()) {
+        Some("continue")
+    } else if FnDef::can_cast(parent_container.kind()) {
+        Some("return;")
+    } else {
+        None
+    }?;
+
+    let cursor_position = ctx.frange.range.start();
+
+    ctx.add_action(AssistId("convert_to_guarded_return"), "convert to guarded return", |edit| {
+        let new_if_expr = format!(
+            "if !{} {{\n    {}\n}}\n{}",
+            expr.syntax().text(),
+            early_expression,
+            then_block.statements().map(|s| s.syntax().text()).join("\n")
```

Comment:
Hm, there might be other useful things in the `then_block`, like comments and local items, and we shouldn't lose those. Not sure what's the best way to handle this: we can copy-paste the text of the block, or we can copy all children_or_token. I'd say we want the latter long-term, but the former might be easier to do at the moment.